Batteries/Data/List/Lemmas.lean

```diff
@@ -346,50 +346,136 @@ theorem disjoint_take_drop : ∀ {l : List α}, l.Nodup → m ≤ n → Disjoint
 
 /-! ### Chain -/
 
-attribute [simp] Chain.nil
+@[simp]
+protected theorem IsChain.nil {R : α → α → Prop} : IsChain R [] := trivial
+
+@[simp]
+protected theorem IsChain.singleton {R : α → α → Prop} : IsChain R [a] := trivial
 
 @[simp]
-theorem chain_cons {a b : α} {l : List α} : Chain R a (b :: l) ↔ R a b ∧ Chain R b l :=
-  ⟨fun p => by cases p with | cons n p => exact ⟨n, p⟩,
-   fun ⟨n, p⟩ => p.cons n⟩
+theorem isChain_cons_cons {a b : α} {l : List α} :
+    IsChain R (a :: b :: l) ↔ R a b ∧ IsChain R (b :: l) :=
+  .rfl
+
+theorem IsChain.cons (h₁ : R a b) (h₂ : IsChain R (b :: l)) : IsChain R (a :: b :: l) :=
+  isChain_cons_cons.mpr ⟨h₁, h₂⟩
+
+theorem IsChain.rel (h : IsChain R (a :: b :: l)) : R a b := (isChain_cons_cons.mp h).1
+
+protected theorem IsChain.tail : ∀ {l}, IsChain R l → IsChain R l.tail
+  | [], _ | [_], _ => .nil
+  | _ :: _ :: _, h => (isChain_cons_cons.mp h).2
+
+theorem IsChain.of_cons (h : IsChain R (a :: l)) : IsChain R l := h.tail
+
+theorem isChain_cons_iff_head :
+    IsChain R (a :: l) ↔ (∀ h : l ≠ [], R a (l.head h)) ∧ l.IsChain R := by
+  cases l <;> simp
+
+theorem isChain_cons_iff_head? : IsChain R (a :: l) ↔ (∀ b ∈ l.head?, R a b) ∧ l.IsChain R := by
+  cases l <;> simp
+
+/-- Recursion principle for `List.IsChain. -/
+@[elab_as_elim]
+protected def IsChain.rec {p : ∀ l : List α, IsChain R l → Sort u}
```

Comment:
Please use `motive` instead of `p`.

Reply:
Maybe add `recOn` for convenience? I'm not sure if `casesOn` would be useful but it doesn't hurt to have it.